Catch JsonIgnore early when parsing annotated resource classes #350

Ramblurr opened the pull request:
See #350

I'm not actually 100% happy with this solution, but maybe it will have to do for now.

With this commit an `IllegalStateException` is no longer thrown when parsing a resource class with a parameterized field type.

However, the parameterized type still appears in the ResourceInformation for the type, because the JsonIgnore is on the getter not the field itself. I think Katharsis should respect `@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.NONE)` if possible.

What do you think @remmeier should that parameterized type show up at all in the  ResourceInformation?

Perhaps we can open another ticket to track that issue.

Ramblurr commented:
@remmeier  @chb0github  Closed the old PR and created this new one against master

Ramblurr commented:
@remmeier bump. This needs a review again.

chb0github commented on katharsis-core/src/main/java/io/katharsis/core/internal/resource/AnnotationResourceInformationBuilder.java:
```diff
@@ -137,7 +138,7 @@ public static String getResourceType(Type genericType, ResourceInformationBuilde
 			AnnotatedResourceField resourceField = new AnnotatedResourceField(jsonName, underlyingName, getter.getReturnType(), getter.getGenericReturnType(), oppositeResourceType, annotations);
-			if (Modifier.isStatic(getter.getModifiers())) {
+			if (Modifier.isStatic(getter.getModifiers()) || resourceField.isAnnotationPresent(JsonIgnore.class)) {
 				fieldWrappers.add(new ResourceFieldWrapper(resourceField, true));
```
this if/else statement could be entirely replaced by simply substituting the predicated for the parameter that's different between the branches. 

I don't think you wrote all the surrounding code, but reducing complexity when/where you can should always be in your mind.

chb0github commented on katharsis-core/src/main/java/io/katharsis/core/internal/resource/AnnotationResourceInformationBuilder.java:
```diff
@@ -110,7 +110,8 @@ public String getResourceType(Class<?> resourceClass) {
-			if (Modifier.isTransient(field.getModifiers()) || Modifier.isStatic(field.getModifiers())) {
+			if (Modifier.isTransient(field.getModifiers()) || Modifier.isStatic(field.getModifiers())
+                    || resourceField.isAnnotationPresent(JsonIgnore.class)) {
 				fieldWrappers.add(new ResourceFieldWrapper(resourceField, true));
```
the only difference between these 2 branches is that lone boolean. This whole `if` could be swapped out

Ramblurr commented:
@chb0github Done! Thanks for pointing that out, that if clause was a bit excessive. 

I've rebased on master and squashed the fix into a single commit.

chb0github commented:
Hmm, no obvious reason for the build to be stuck. Let's see how it looks tomorrow

chb0github commented:
Hmm, there is a problem with NPM

```
travis_time:end:100c7aa5:start=1490006806696555177,finish=1490006806930178354,duration=233623177
[0Ktravis_fold:end:before_install.8
[0Ktravis_fold:start:before_install.9
[0Ktravis_time:start:2f154ce3
[0K$ node --version
v6.10.0

travis_time:end:2f154ce3:start=1490006806935577921,finish=1490006806942105459,duration=6527538
[0Ktravis_fold:end:before_install.9
[0Ktravis_fold:start:before_install.10
[0Ktravis_time:start:0e8aed9c
[0K$ sudo npm install -g @angular/cli@latest
[37;40mnpm[0m [0m[31;40mERR![0m[35m[0m Linux 4.4.0-51-generic
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35margv[0m "/usr/bin/nodejs" "/usr/bin/npm" "install" "-g" "@angular/cli@latest"
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mnode[0m v7.7.3
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mnpm [0m v4.1.2
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mcode[0m ECONNRESET
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35merrno[0m ECONNRESET
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35msyscall[0m read
[0m
[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mnetwork[0m read ECONNRESET
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mnetwork[0m This is most likely not a problem with npm itself
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mnetwork[0m and is related to network connectivity.
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mnetwork[0m In most cases you are behind a proxy or have bad network settings.
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mnetwork[0m 
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mnetwork[0m If you are behind a proxy, please make sure that the
[0m[37;40mnpm[0m [0m[31;40mERR![0m [0m[35mnetwork[0m 'proxy' config is set properly.  See: 'npm help config'
[0m
[37;40mnpm[0m [0m[31;40mERR![0m[35m[0m Please include the following file with any support request:
[0m[37;40mnpm[0m [0m[31;40mERR![0m[35m[0m     /home/travis/build/katharsis-project/katharsis-framework/npm-debug.log
[0m
travis_time:end:0e8aed9c:start=1490006806947265737,finish=1490007086423734359,duration=279476468622
[0K
[31;1mThe command "sudo npm install -g @angular/cli@latest" failed and exited with 1 during .[0m
```

@remmeier - what's the purpose of `sudo npm install -g @angular/cli@latest` ? If you're going to build dependencies for a node module it should be built in the directory itself. This is what's failing

chb0github commented:
Without an update on this PR and the need/use of NPM in the build script I guess it should be closed.

Ramblurr commented:
@chb0github I'm confused why should this be closed?

1. I made the requested changes
2. I didn't have anything to do with npm in the build script, that is unrelated to the commit in this PR. 

Moreover, this fixes a critical bug that prevents us from updating to 3.0

Ramblurr commented:
This is failing on travis, but locally a completely clean test suite passes with 0 errors. 

If you see the error `io.katharsis.errorhandling.exception.ResourceNotFoundException: failed to find resource with id 1` and think it is failing the build, in fact it is a red herring.

See [this build](https://travis-ci.org/katharsis-project/katharsis-framework/builds/217928646) by @chb0github , look in the log and you will also see the same exception thrown, but it didn't fail the build.

Something is broken with regards to travis.

Sandmania commented:
If it helps solving the build problem, the same test failure happens here https://github.com/katharsis-project/katharsis-framework/pull/398

Locally katharsis-client tests work just fine.

chb0github commented:
I didn't close it. And, I got the build working again.

chb0github commented:
Ok, GH won't let me merge in master again. Not sure WTF. @Ramblurr can you merge master in again and push? If it goes green I will merge in. It's a pita

Ramblurr commented:
Moved to https://github.com/katharsis-project/katharsis-framework/pull/397
